File: client/listonit/lib/features/items/domain/item.dart
import 'package:equatable/equatable.dart';

class Item extends Equatable {
  final String id;
  final String listId;
  final String name;
  final int quantity;
  final String? unit;
  final String? note;
  final bool isChecked;
  final DateTime? checkedAt;
  final String? checkedBy;
  final int sortIndex;
  final String createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isLocal;

  const Item({
    required this.id,
    required this.listId,
    required this.name,
    this.quantity = 1,
    this.unit,
    this.note,
    this.isChecked = false,
    this.checkedAt,
    this.checkedBy,
    this.sortIndex = 0,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
    this.isLocal = false,
  });

  factory Item.fromJson(Map<String, dynamic> json) {
    return Item(
      id: json['id'] as String,
      listId: json['list_id'] as String,
      name: json['name'] as String,
      quantity: json['quantity'] as int? ?? 1,
      unit: json['unit'] as String?,
      note: json['note'] as String?,
      isChecked: json['is_checked'] as bool? ?? false,
      checkedAt: json['checked_at'] != null
          ? DateTime.parse(json['checked_at'] as String)
          : null,
      checkedBy: json['checked_by'] as String?,
      sortIndex: json['sort_index'] as int? ?? 0,
      createdBy: json['created_by'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'list_id': listId,
      'name': name,
      'quantity': quantity,
      'unit': unit,
      'note': note,
      'is_checked': isChecked,
      'checked_at': checkedAt?.toIso8601String(),
      'checked_by': checkedBy,
      'sort_index': sortIndex,
      'created_by': createdBy,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Item copyWith({
    String? id,
    String? listId,
    String? name,
    int? quantity,
    String? unit,
    String? note,
    bool? isChecked,
    DateTime? checkedAt,
    String? checkedBy,
    int? sortIndex,
    String? createdBy,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isLocal,
  }) {
    return Item(
      id: id ?? this.id,
      listId: listId ?? this.listId,
      name: name ?? this.name,
      quantity: quantity ?? this.quantity,
      unit: unit ?? this.unit,
      note: note ?? this.note,
      isChecked: isChecked ?? this.isChecked,
      checkedAt: checkedAt ?? this.checkedAt,
      checkedBy: checkedBy ?? this.checkedBy,
      sortIndex: sortIndex ?? this.sortIndex,
      createdBy: createdBy ?? this.createdBy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isLocal: isLocal ?? this.isLocal,
    );
  }

  @override
  List<Object?> get props => [
        id,
        listId,
        name,
        quantity,
        unit,
        note,
        isChecked,
        checkedAt,
        checkedBy,
        sortIndex,
        createdBy,
        createdAt,
        updatedAt,
        isLocal,
      ];
}

File: backend/models/item.py
import uuid
from datetime import datetime
from sqlalchemy import String, DateTime, Boolean, Integer, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from database import Base


class Item(Base):
    """Shopping list item model."""

    __tablename__ = "items"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    list_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("shopping_lists.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(200))
    quantity: Mapped[int] = mapped_column(Integer, default=1)
    unit: Mapped[str | None] = mapped_column(String(20), nullable=True)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_checked: Mapped[bool] = mapped_column(Boolean, default=False)
    checked_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    checked_by: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("users.id"), nullable=True
    )
    sort_index: Mapped[int] = mapped_column(Integer, default=0)
    created_by: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id"), index=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    # Relationships
    shopping_list: Mapped["ShoppingList"] = relationship(
        "ShoppingList", back_populates="items"
    )
    creator: Mapped["User"] = relationship(
        "User", foreign_keys=[created_by], backref="created_items"
    )
    checker: Mapped["User"] = relationship(
        "User", foreign_keys=[checked_by], backref="checked_items"
    )
